internal/contour/endpointstranslator.go

```diff
@@ -26,30 +28,295 @@ import (
+			if port.Protocol != p.Protocol && p.Protocol != v1.ProtocolTCP {
+				// NOTE: we only support "TCP", which is the default.
+				continue
+			}
+
+			// If the port isn't named, it must be the
+			// only Service port, so it's a match by
+			// definition. Otherwise, only take endpoint
+			// ports that match the service port name.
+			if port.Name != "" && port.Name != p.Name {
+				continue
+			}
+
+			// If we matched this port, collect Envoy endpoints for all the ready addresses.
+			addresses := append([]v1.EndpointAddress{}, s.Addresses...) // Shallow copy.
+			sort.Slice(addresses, func(i, j int) bool { return addresses[i].IP < addresses[j].IP })
+
+			for _, a := range addresses {
+				addr := envoy.SocketAddress(a.IP, int(p.Port))
+				lb = append(lb, envoy.LBEndpoint(addr))
+			}
+		}
+	}
+
+	return lb
+}
+
+// EndpointsCache is a cache of Endpoint and ServiceCluster objects.
+type EndpointsCache struct {
+	mu sync.Mutex // Protects all fields.
+
+	// Slice of stale clusters. A stale cluster is one that
+	// needs to be recalculated. Clusters can be added to the stale
+	// slice due to changes in Endpoints or due to a DAG rebuild.
+	stale []*dag.ServiceCluster
+
+	// Index of ServiceClusters. ServiceClusters are indexed
+	// by the name of their Kubernetes Services. This makes it
+	// easy to determine which Endpoints affect which ServiceCluster.
+	services map[types.NamespacedName][]*dag.ServiceCluster
+
+	// Cache of endpoints, indexed by name.
+	endpoints map[types.NamespacedName]*v1.Endpoints
+}
+
+// Recalculate regenerates all the ClusterLoadAssignments from the
+// cached Endpoints and stale ServiceClusters. A ClusterLoadAssignment
+// will be generated for every stale ServerCluster, however, if there
+// are no endpoints for the Services in the ServiceCluster, the
+// ClusterLoadAssignment wil be empty.
+func (c *EndpointsCache) Recalculate() map[string]*v2.ClusterLoadAssignment {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	assignments := map[string]*v2.ClusterLoadAssignment{}
+	for _, cluster := range c.stale {
+		// Clusters can be in the stale list multiple times;
+		// skip to avoid duplicate recalculations.
+		if _, ok := assignments[cluster.ClusterName]; ok {
+			continue
+		}
+
+		cla := v2.ClusterLoadAssignment{
+			ClusterName: cluster.ClusterName,
+			Endpoints:   nil,
+			Policy:      nil,
+		}
+
+		// Look up each service, and if we have endpoints for that service,
+		// attach them as a new LocalityEndpoints resource.
+		for _, w := range cluster.Services {
+			n := types.NamespacedName{Namespace: w.ServiceNamespace, Name: w.ServiceName}
+			if lb := RecalculateEndpoints(w.ServicePort, c.endpoints[n]); lb != nil {
+				cla.Endpoints = append(
+					cla.Endpoints,
+					&LocalityEndpoints{
+						LbEndpoints:         lb,
+						LoadBalancingWeight: protobuf.UInt32(w.Weight),
+					},
+				)
+			}
+		}
+
+		assignments[cla.ClusterName] = &cla
+		log.Printf("currently %d clusters", len(assignments))
```

Comment:
This should be a debug level and use the logger to print properly. 

![image](https://user-images.githubusercontent.com/1048184/91233844-0948be80-e700-11ea-8b4c-c1a5813898bd.png)